src/rdkafka_lz4.c

```diff
@@ -0,0 +1,430 @@
+                           inlen);
+                return RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
+        }
+
+        of = 4; /* past magic */
+        FLG = inbuf[of++];
+        of++; /* BD */
+
+        if ((FLG >> 3) & 1) /* contentSize */
+                of += 8;
+
+        if (of >= inlen) {
+                rd_rkb_dbg(rkb, BROKER,  "LZ4FIXUP",
+                           "Unable to fix-up legacy LZ4 framing "
+                           "(%"PRIusz" bytes): requires %"PRIusz" bytes",
+                           inlen, of);
+                return RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
+        }
+
+        /* Header hash code */
+        HC = inbuf[of];
+
+        /* Calculate correct header hash code */
+        correct_HC = (XXH32(inbuf+4, of-4, 0) >> 8) & 0xff;
+
+        if (HC != correct_HC)
+                inbuf[of] = correct_HC;
+
+        return RD_KAFKA_RESP_ERR_NO_ERROR;
+}
+
+
+/**
+ * Reverse of fix-up: break LZ4 framing caused to be compatbile with with
+ * buggy Kafka client / broker.
+ *
+ * NOTE: This modifies 'outbuf'.
+ *
+ * Returns an error on failure to recognize format (nothing modified),
+ * else NO_ERROR.
+ */
+static rd_kafka_resp_err_t
+rd_kafka_lz4_compress_break_framing (rd_kafka_broker_t *rkb,
+                                     char *outbuf, size_t outlen) {
+        static const char magic[4] = { 0x04, 0x22, 0x4d, 0x18 };
+        uint8_t FLG, HC, bad_HC;
+        size_t of = 4;
+
+        /* Format is:
+         *    int32_t magic;
+         *    int8_t_ FLG;
+         *    int8_t  BD;
+         *  [ int64_t contentSize; ]
+         *    int8_t  HC;
+         */
+        if (outlen < 4+3 || memcmp(outbuf, magic, 4)) {
+                rd_rkb_dbg(rkb, BROKER,  "LZ4FIXDOWN",
+                           "Unable to break legacy LZ4 framing "
+                           "(%"PRIusz" bytes): invalid length or magic value",
+                           outlen);
+                return RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
+        }
+
+        of = 4; /* past magic */
+        FLG = outbuf[of++];
+        of++; /* BD */
+
+        if ((FLG >> 3) & 1) /* contentSize */
+                of += 8;
+
+        if (of >= outlen) {
+                rd_rkb_dbg(rkb, BROKER,  "LZ4FIXUP",
+                           "Unable to break legacy LZ4 framing "
+                           "(%"PRIusz" bytes): requires %"PRIusz" bytes",
+                           outlen, of);
+                return RD_KAFKA_RESP_ERR__BAD_COMPRESSION;
+        }
+
+        /* Header hash code */
+        HC = outbuf[of];
+
+        /* Calculate bad header hash code (include magic) */
+        bad_HC = (XXH32(outbuf, of, 0) >> 8) & 0xff;
+
+        if (HC != bad_HC)
+                outbuf[of] = bad_HC;
+
+        return RD_KAFKA_RESP_ERR_NO_ERROR;
+}
+
+
+
+/**
+ * @brief Decompress LZ4F (framed) data.
+ *        Kafka broker versions <0.10.0.0 breaks LZ4 framing checksum, if
+ *        \p proper_hc we assume the checksum is okay (broker version >=0.10.0)
```

Comment:
Nit: it should probably talk about message format versions instead of broker versions, no?